pkg/sync/parallel_syncer.go

```diff
@@ -0,0 +1,1706 @@
+package sync
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/grpc-ecosystem/go-grpc-middleware/logging/zap/ctxzap"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.uber.org/zap"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	v2 "github.com/conductorone/baton-sdk/pb/c1/connector/v2"
+	reader_v2 "github.com/conductorone/baton-sdk/pb/c1/reader/v2"
+	"github.com/conductorone/baton-sdk/pkg/annotations"
+)
+
+var _ Syncer = (*parallelSyncer)(nil)
+
+var parallelTracer = otel.Tracer("baton-sdk/parallel-sync")
+
+// StateInterface defines the minimal interface needed by helper methods
+// This allows helper methods to work with either the sequential syncer's state machine
+// or the parallel syncer's local state context
+type StateInterface interface {
+	PageToken(ctx context.Context) string
+	NextPage(ctx context.Context, pageToken string) error
+	SetNeedsExpansion()
+	SetHasExternalResourcesGrants()
+	ShouldFetchRelatedResources() bool
+}
+
+// ActionDecision represents the decision made by a helper method
+// This allows the caller to decide how to handle the result
+type ActionDecision struct {
+	ShouldContinue       bool   // Whether to continue processing (e.g., more pages)
+	NextPageToken        string // Page token for next page, if applicable
+	Action               string // What action to take: "next_page", "finish", "continue"
+	NeedsExpansion       bool   // Whether grant expansion is needed
+	HasExternalResources bool   // Whether external resources were found
+	ShouldFetchRelated   bool   // Whether related resources should be fetched
+}
+
+// LocalStateContext provides local state management for parallel syncer
+// This implements StateInterface without sharing the global state machine
+type LocalStateContext struct {
+	resourceID           *v2.ResourceId
+	pageToken            string
+	needsExpansion       bool
+	hasExternalResources bool
+	shouldFetchRelated   bool
+}
+
+// NewLocalStateContext creates a new local state context for a resource
+func NewLocalStateContext(resourceID *v2.ResourceId) *LocalStateContext {
+	return &LocalStateContext{
+		resourceID:           resourceID,
+		pageToken:            "",
+		needsExpansion:       false,
+		hasExternalResources: false,
+		shouldFetchRelated:   true, // Default to true for backward compatibility
```

Comment:
I think this defaults to false in sequential syncer. It's only turned on for incremental sync. Also it only works for resource types that implement Get() for incremental sync.

If we're hitting a rate limited API, getting resources one at a time can really hurt performance, so I think this should be off except for incremental syncs.